File: apps/worker/src/db/items_repo.ts
import type { NormalizedEntry } from '../ingest/types';

/**
 * Upsert a batch of normalized entries into the items table.
 * Uses INSERT OR IGNORE with item_key as the dedup key (sha256 of normalizedUrl).
 * item_id is set equal to item_key for deterministic, idempotent inserts.
 *
 * Returns the count of rows that were actually inserted (new items only).
 */
export async function upsertItems(
  db: D1Database,
  entries: NormalizedEntry[],
  sourceId: string,
): Promise<{ found: number; inserted: number; newKeys: string[] }> {
  if (entries.length === 0) return { found: 0, inserted: 0, newKeys: [] };

  const now = new Date().toISOString();

  const stmts = entries.map(e =>
    db
      .prepare(
        `INSERT OR IGNORE INTO items
           (item_id, source_id, source_url, normalized_url, item_key,
            title_he, published_at, updated_at, date_confidence,
            snippet_he, title_hash, ingested_at, status)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'new')`,
      )
      .bind(
        e.itemKey,        // item_id  = item_key (deterministic, idempotent)
        sourceId,
        e.sourceUrl,
        e.normalizedUrl,
        e.itemKey,        // item_key = sha256(normalizedUrl)
        e.titleHe,
        e.publishedAt,
        e.publishedAt,    // updated_at mirrors published_at on insert
        e.dateConfidence,
        e.snippetHe,
        e.titleHash,
        now,
      ),
  );

  const results = await db.batch(stmts);
  const newKeys: string[] = [];
  for (let i = 0; i < results.length; i++) {
    if ((results[i]?.meta?.changes ?? 0) > 0) {
      newKeys.push(entries[i]!.itemKey);
    }
  }
  return { found: entries.length, inserted: newKeys.length, newKeys };
}

File: apps/local-engine/db/repos/items_repo.py
"""db/repos/items_repo.py — Item upsert (INSERT OR IGNORE)."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime

import aiosqlite

from ingest.rss import NormalizedEntry


def _now_iso() -> str:
    return datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"


@dataclass
class UpsertResult:
    found: int              # total items attempted
    inserted: int           # actually new
    new_keys: set[str] = field(default_factory=set)  # item_keys that were new


async def upsert_items(
    db: aiosqlite.Connection,
    entries: list[NormalizedEntry],
    source_id: str,
) -> UpsertResult:
    """INSERT OR IGNORE entries into the items table.

    Idempotent: re-running with same entries produces no duplicates.
    Uses cursor.rowcount after each statement to detect actual inserts.
    """
    ingested_at = _now_iso()
    new_keys: set[str] = set()

    for entry in entries:
        async with db.execute(
            """
            INSERT OR IGNORE INTO items (
              item_id, source_id, source_url, normalized_url, item_key,
              title_he, published_at, date_confidence, snippet_he,
              title_hash, ingested_at, status, enclosure_url
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'new', ?)
            """,
            (
                entry.item_key,   # item_id = item_key (deterministic)
                source_id,
                entry.source_url,
                entry.normalized_url,
                entry.item_key,
                entry.title_he,
                entry.published_at,
                entry.date_confidence,
                entry.snippet_he,
                entry.title_hash,
                ingested_at,
                entry.enclosure_url,
            ),
        ) as cursor:
            if cursor.rowcount > 0:
                new_keys.add(entry.item_key)

    await db.commit()

    return UpsertResult(
        found=len(entries),
        inserted=len(new_keys),
        new_keys=new_keys,
    )
